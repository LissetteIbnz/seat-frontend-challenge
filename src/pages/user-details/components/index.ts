export * from "./form.component";
export * from "./input-control.component";
