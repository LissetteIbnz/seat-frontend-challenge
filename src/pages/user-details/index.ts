export * from "./user-details.container";
