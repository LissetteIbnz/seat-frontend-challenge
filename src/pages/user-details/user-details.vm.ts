export type UserID = string;

export type User = {
  id: UserID;
  firstName: string;
  lastName: string;
  email: string;
};

export const createEmptyUser = (): User => ({
  id: "",
  email: "",
  firstName: "",
  lastName: "",
});
