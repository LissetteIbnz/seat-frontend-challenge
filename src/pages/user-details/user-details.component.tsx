import { Box, Button, Text } from "@chakra-ui/react";
import { useFormik } from "formik";
import { InputControl, Form } from "./components";
import { validateForm } from "./user-details.validations";
import { User } from "./user-details.vm";

interface Props {
  isEditMode: boolean;
  onCancel: VoidFunction;
  onSave: (user: User) => void;
  user: User;
}

export const UserDetailsComponent = ({ user, isEditMode, onCancel, onSave }: Props) => {
  const { errors, isSubmitting, isValid, values, touched, handleBlur, handleSubmit, handleChange } =
    useFormik({
      initialValues: user,
      validate: validateForm,
      onSubmit: (values) => {
        onSave(values);
      },
      enableReinitialize: true,
    });

  return (
    <Form onSubmit={handleSubmit}>
      {isEditMode && <Text mb={3}>ID: {user.id}</Text>}
      <InputControl<User>
        id="firstName"
        label="First name"
        isInvalid={Boolean(errors.firstName) && touched.firstName}
        value={values.firstName}
        onChange={handleChange}
        onBlur={handleBlur}
        error={errors.firstName}
      />
      <InputControl<User>
        id="lastName"
        label="Last name"
        isInvalid={Boolean(errors.lastName) && touched.lastName}
        value={values.lastName}
        onChange={handleChange}
        onBlur={handleBlur}
        error={errors.lastName}
      />
      <InputControl<User>
        id="email"
        label="Email"
        isInvalid={Boolean(errors.email) && touched.email}
        value={values.email}
        onChange={handleChange}
        onBlur={handleBlur}
        error={errors.email}
      />

      <Box mt={4} alignSelf="flex-end">
        <Button variant="ghost" mr={2} type="button" onClick={onCancel}>
          Cancel
        </Button>
        <Button colorScheme="teal" disabled={!isValid} isLoading={isSubmitting} type="submit">
          Submit
        </Button>
      </Box>
    </Form>
  );
};
