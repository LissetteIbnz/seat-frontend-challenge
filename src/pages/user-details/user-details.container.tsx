import * as React from "react";
import { useNavigate, useParams } from "react-router-dom";
import { useToast } from "@chakra-ui/react";
import { getErrorMessage } from "common/helpers";
import { routes } from "core/router";
import { UserDetailsComponent } from "./user-details.component";
import { isValidId, getUserById, createUser } from "./user-details.service";
import { createEmptyUser, User } from "./user-details.vm";

export const UserDetailsContainer = () => {
  const { id } = useParams();
  const navigate = useNavigate();
  const toast = useToast({
    position: "top-right",
  });

  const [user, setUser] = React.useState<User>(createEmptyUser());
  const [isEditMode, setIsEditMode] = React.useState(false);

  const onLoadUser = async (id: string) => {
    try {
      const vmUser = await getUserById(id);
      setUser(vmUser);
    } catch (error) {
      const message = getErrorMessage(error);
      toast({
        description: message,
        status: "error",
      });
    }
  };

  React.useEffect(() => {
    const isValid = isValidId(id);
    setIsEditMode(isValid);
    if (isValid) {
      onLoadUser(id);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleSave = async (newUser: User) => {
    try {
      await createUser(newUser);
      toast({
        description: "User created successfully",
        status: "success",
      });
      goBack();
    } catch (error) {
      toast({
        description: getErrorMessage(error),
        status: "error",
      });
    }
  };

  const goBack = () => {
    navigate(routes.users);
  };

  return (
    <UserDetailsComponent
      user={user}
      isEditMode={isEditMode}
      onCancel={goBack}
      onSave={handleSave}
    />
  );
};
