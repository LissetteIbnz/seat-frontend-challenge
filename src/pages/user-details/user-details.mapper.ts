import { isUndefinedOrNull } from "common/helpers";
import * as AM from "infra/dto";
import * as VM from "./user-details.vm";

export const mapUserFromApiToVm = (user: AM.UserDTO): VM.User =>
  isUndefinedOrNull(user)
    ? VM.createEmptyUser()
    : {
        email: user.email,
        firstName: user.first_name,
        lastName: user.last_name,
        id: user.id.toString(),
      };

export const mapUserFromVmToApi = (user: VM.User): AM.UserDTO => ({
  email: user.email,
  first_name: user.firstName,
  last_name: user.lastName,
  id: parseInt(user.id),
});
