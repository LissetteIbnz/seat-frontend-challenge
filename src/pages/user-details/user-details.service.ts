import { DEFAULT_USER_ID } from "common/models";
import { userRepository } from "infra/repositories";
import { mapUserFromApiToVm, mapUserFromVmToApi } from "./user-details.mapper";
import * as VM from "./user-details.vm";

export const getUserById = async (userId: string): Promise<VM.User> => {
  const apiUser = await userRepository.getById(userId);
  const vmUser = mapUserFromApiToVm(apiUser);
  return vmUser;
};

export const createUser = async (user: VM.User): Promise<VM.User> => {
  const mapUser = mapUserFromVmToApi(user);
  const apiUser = await userRepository.create(mapUser);
  const vmUser = mapUserFromApiToVm(apiUser);
  return vmUser;
};

export const isValidId = (id: string): boolean => {
  return id !== DEFAULT_USER_ID && Boolean(id);
};
