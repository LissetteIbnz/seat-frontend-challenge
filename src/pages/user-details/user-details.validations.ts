import { User } from "./user-details.vm";

export const validateForm = (form: User): Partial<Record<keyof User, string>> => {
  const errors: Partial<Record<keyof User, string>> = {};

  if (!form?.email) {
    errors.email = "Email is required";
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(form.email)) {
    errors.email = "Invalid email address";
  }

  if (!form?.firstName) {
    errors.firstName = "First Name is required";
  }

  if (!form?.lastName) {
    errors.lastName = "Last Name is required";
  }

  return errors;
};
