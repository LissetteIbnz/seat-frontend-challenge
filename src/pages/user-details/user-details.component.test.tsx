import { waitFor } from "@testing-library/react";
import { render, screen, changeInput, fireEvent } from "test-utils";
import { UserDetailsComponent } from "./user-details.component";

type Props = React.ComponentProps<typeof UserDetailsComponent>;

describe("pages/user-details/UserDetailsComponent", () => {
  let props: Props;

  beforeEach(() => {
    props = {
      user: {
        email: "",
        firstName: "",
        lastName: "",
        id: "",
      },
      isEditMode: false,
      onCancel: jest.fn(),
      onSave: jest.fn(),
    };
  });

  it("should render correctly", () => {
    const { container } = render(<UserDetailsComponent {...props} />);
    expect(container).toMatchSnapshot();
  });

  describe("given isEditMode is false", () => {
    beforeEach(() => {
      props.isEditMode = true;
    });

    it("should call onSave", async () => {
      render(<UserDetailsComponent {...props} />);

      const submitButton = screen.getByRole("button", { name: /submit/i });
      expect(props.onSave).toHaveBeenCalledTimes(0);

      const firstNameInput = screen.getByLabelText<HTMLInputElement>(/first name/i);
      const lastNameInput = screen.getByLabelText<HTMLInputElement>(/last name/i);
      const emailInput = screen.getByLabelText<HTMLInputElement>(/email/i);

      changeInput(firstNameInput, "John");
      changeInput(lastNameInput, "Doe");
      changeInput(emailInput, "wrong-email");

      expect(submitButton).not.toBeDisabled();
      fireEvent.click(submitButton);

      changeInput(emailInput, "valid-email@email.com");
      fireEvent.click(submitButton);

      await waitFor(() => {
        expect(props.onSave).toHaveBeenCalledTimes(1);
      });
    });

    it("should call onCancel", async () => {
      render(<UserDetailsComponent {...props} />);

      const cancelButton = screen.getByRole("button", { name: /cancel/i });
      expect(props.onCancel).toHaveBeenCalledTimes(0);

      fireEvent.click(cancelButton);

      await waitFor(() => {
        expect(props.onCancel).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe("given isEditMode is true", () => {
    beforeEach(() => {
      props.isEditMode = true;
      props.user = {
        email: "irrelevant@email.com",
        firstName: "John",
        lastName: "Doe",
        id: "irrelevant-id",
      };
    });

    it("should call onSave with new changes", async () => {
      render(<UserDetailsComponent {...props} />);

      const submitButton = screen.getByRole("button", { name: /submit/i });
      expect(props.onSave).toHaveBeenCalledTimes(0);

      const firstNameInput = screen.getByLabelText<HTMLInputElement>(/first name/i);
      const lastNameInput = screen.getByLabelText<HTMLInputElement>(/last name/i);
      const emailInput = screen.getByLabelText<HTMLInputElement>(/email/i);

      changeInput(firstNameInput, "Updated John");
      changeInput(lastNameInput, "Updated Doe");
      changeInput(emailInput, "update-email@email.com");

      expect(submitButton).not.toBeDisabled();
      fireEvent.click(submitButton);

      await waitFor(() => {
        expect(props.onSave).toHaveBeenCalledWith(
          expect.objectContaining({
            email: "update-email@email.com",
            firstName: "Updated John",
            lastName: "Updated Doe",
          }),
        );
      });
    });
  });
});
