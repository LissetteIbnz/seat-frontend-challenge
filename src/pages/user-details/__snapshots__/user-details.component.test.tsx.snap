// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/user-details/UserDetailsComponent should render correctly 1`] = `
<div>
  <form
    class="css-4euc2w"
  >
    <div
      class="chakra-form-control css-rj5r62"
      role="group"
    >
      <label
        class="chakra-form__label css-2gx1h6"
        for="firstName"
        id="field-1-label"
      >
        First name
      </label>
      <input
        class="chakra-input css-1c6j008"
        id="firstName"
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-rj5r62"
      role="group"
    >
      <label
        class="chakra-form__label css-2gx1h6"
        for="lastName"
        id="field-2-label"
      >
        Last name
      </label>
      <input
        class="chakra-input css-1c6j008"
        id="lastName"
        value=""
      />
    </div>
    <div
      class="chakra-form-control css-rj5r62"
      role="group"
    >
      <label
        class="chakra-form__label css-2gx1h6"
        for="email"
        id="field-3-label"
      >
        Email
      </label>
      <input
        class="chakra-input css-1c6j008"
        id="email"
        value=""
      />
    </div>
    <div
      class="css-n3wvi9"
    >
      <button
        class="chakra-button css-2rnqoq"
        type="button"
      >
        Cancel
      </button>
      <button
        class="chakra-button css-f6n2t5"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
